### ISSUE TITLE
Initial setup for all tests

### ISSUE THREAD

**Issue by spawnia** [paratestphp/paratest#656]
## Motivation

In order to speed up my tests, they perform a one-time setup of a shared cache. This cache has to be written exactly once and is then only read. I want it to be refreshed when re-running tests - its contents depend on the application code.

When using plain PHPUnit, I could solve this by remembering if the setup ran in a static variable:

```php
static bool $initialized = false;

public function setUp() {
    if (! self::$initialized) {
         self::initialize();
         self::$initialized = true;
    }
}
```

## Problem

However, when using parallel testing, the initialization step happens once for each parallel job. I tested this by making `initialize()` append to a file, and found it was indeed happening multiple times.

I though about separating state between workers through `TEST_TOKEN`, but in this case it does not make sense. It would complicate the implementation of the cache mechanism, and it would perform unnecessary work - the initialization only needs to happen exactly once for all tests. We are using this mechanism to create separate test databases though, they are mutable and thus should be separated per process.

## Proposed solution

It would be nice if Paratest offered a way to register an initialization step that runs exactly once before all tests. Per-process initialization is useful for process-specific tasks  such as initializing PHP code or setting up shared mutable state, but that does not cover every use case. It is desirable to initialize shared immutable caches once.

This way of initialization is possible when using plain PHPUnit, since in that case the semantics of once-for-all-tests are the same as once-per-process. Paratest only offers initialization once-per-process, but lacks the ability to have a once-for-all-tests initialization.

**Comment by Slamdunk** [comment]
There already is a way to achieve this: initialize the cache only when `TEST_TOKEN` is **not** set.

**Comment by spawnia** [comment]
Thanks for the quick response. I am not sure I understand your suggestion entirely. How would that work if I am running my testsuite with Paratest? As far as I understand, every process will have `TEST_TOKEN` populated.

Let me rephrase what I can observe when running my tests:
- `vendor/bin/phpunit`: the initialization step described in my code example happens exactly once, blocking and before all tests.
- `vendor/bin/paratest`: the initialization happens once per test process, in parallel. This leads to all kinds of issues, since that step writes a cache to a shared location.

I would like the cache to be intialized exactly once, blocking and before all tests in both cases. What you are suggesting would lead to no cache being initialized at all when running with `vendor/bin/paratest`. This causes a significant slowdown of my tests, to the point where the parallel execution is no longer faster.

|  | PHPUnit | Paratest|
|---|---|---|
| Initialize per process | ✔️ bootstrap, checked through static variable | ✔️ TEST_TOKEN |
| Initialize for all | ✔️ same as per process due to single threaded-ness | ❌ not possible |

**Comment by Slamdunk** [comment]
[PHPUnit bootstrap](https://phpunit.readthedocs.io/en/9.5/configuration.html#the-bootstrap-attribute) gets loaded with Paratest both by Paratest and by each parallel process.

So, if you put in your bootstrap:
```php
if (false === getenv('TEST_TOKEN')) {
    initMyCache();
}
```
`initMyCache()` will be called as you expected:
1. one time only is you run bare PHPUnit
2. also one time only if you run ParaTest, regardless of the chosen `--runner`

**Comment by spawnia** [comment]
One additional thing. I realize that I could achieve the result I am looking for by externalizing the initialization step and simply doing it before running any tests at all, regardless whether I am using PHPUnit or Paratest. For example, I could modify the command I use to run tests:

```diff
-vendor/bin/phpunit
-vendor/bin/paratest
+php initialize-cache.php && vendor/bin/phpunit
+php initialize-cache.php && vendor/bin/paratest
```

There are several reasons why this is not optimal though:
- the initialization is only necessary for a subset of tests and registered through their `setUp()` method. When I run only a part of my testsuite (e.g. with `--filter`), I might not need it
- running tests through PHPStorm just calls the PHPUnit binary without my custom script

**Comment by spawnia** [comment]
> So, if you put in your bootstrap:

Thanks, that did indeed work.

I can see a few problems with this solution though (similar to https://github.com/paratestphp/paratest/issues/656#issuecomment-1050784121):
- the initialization is only necessary for a subset of tests and registered through their `setUp()` method. When I run only a part of my testsuite (e.g. with `--filter`), I might not need it.
- the initialization step is actually provided by a library I am using as a trait to use in test classes. In order to work, it requires the application to be bootstrapped. That means that I would have to do a bunch of extra work to extract it to something that can run on its own.

I realize those are not major problems and your solution unblocks me to actually use Paratest, but perhaps we can still improve upon it. Does Paratest provide access to any shared mutable state between test workers? A primitive would for that would allow the workers to coordinate. My initial example might look like this:

```php
public function setUp() {
    if (! Paratest::$sharedTestState['initialized'] ?? false) {
         self::initialize();
         Paratest::$sharedTestState['initialized'] = true;
    }
}
```

Or perhaps it could be possible to register a one-time setup through a hook:

```php
public function setUp() {
    Paratest::onceBeforeAll(function () {
         self::initialize();
    });
}
```

If something like that would be theoretically possible, I am willing to provide a PR for it.

**Comment by Slamdunk** [comment]
> Does Paratest provide access to any shared mutable state between test workers?

Yes: the Filesystem.

Your `setUp` should test if a particular file exists:
1. If it doesn't exist, it creates it and acquire a `LOCK_EX` until the cache warmup has finished
2. If it exists, it acquires a `LOCK_SH` and proceed with the test as soon as that lock type is available

See https://www.php.net/manual/en/function.flock.php

Immediately compatible with both PHPUnit and Paratest, no bootstrap required, `--filter` compatible.

Paratest is already complex enough, because its goal is tricky, so no feature is going to be added when achievable in other ways.

**Comment by spawnia** [comment]
Thanks for your comprehensive answer and guidance, I really appreciate it. Would you accept a PR that documents how to deal with such global initialization concerns when I figure it out in my project?

**Comment by Slamdunk** [comment]
Sure :+1:
